File: dotnet/CucumberExpressions/ExpressionFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace CucumberExpressions;

/// <summary>
/// Creates a {@link CucumberExpression} or {@link RegularExpression} from a {@link String}
/// using heuristics. 
/// </summary>
public class ExpressionFactory {

    private static readonly Regex BEGIN_ANCHOR = new("^\\^.*");
    private static readonly Regex END_ANCHOR = new(".*\\$$");
    private static readonly Regex SCRIPT_STYLE_REGEXP = new("^/(?<inner>.*)/$");
    private static readonly Regex PARAMETER_PATTERN = new("((?:\\\\){0,2})\\{([^}]*)\\}");

    private readonly IParameterTypeRegistry parameterTypeRegistry;

    public ExpressionFactory(IParameterTypeRegistry parameterTypeRegistry) {
        this.parameterTypeRegistry = parameterTypeRegistry;
    }

    public Expression createExpression(String expressionString) {
        if (BEGIN_ANCHOR.IsMatch(expressionString) || END_ANCHOR.IsMatch(expressionString)) {
            return createRegularExpressionWithAnchors(expressionString);
        }
        var m = SCRIPT_STYLE_REGEXP.Match(expressionString);
        if (m.Success) {
            return new RegularExpression(new(m.Groups["inner"].Value), parameterTypeRegistry);
        }
        if (IsRegularExpression(expressionString))
            return createRegularExpressionWithAnchors(expressionString);
        return new CucumberExpression(expressionString, parameterTypeRegistry);
    }

    protected virtual bool IsRegularExpression(string expressionString)
    {
        return false; // additional heuristics can be implemented in derived classes
    }

    private RegularExpression createRegularExpressionWithAnchors(String expressionString) {
        try {
            return new RegularExpression(new(expressionString), parameterTypeRegistry);
        } catch (ArgumentException e) {
            if (PARAMETER_PATTERN.IsMatch(expressionString)) {
                throw new CucumberExpressionException("You cannot use anchors (^ or $) in Cucumber Expressions. Please remove them from " + expressionString, e);
            }
            throw new CucumberExpressionException($"Invalid regular expression: '{expressionString}'", e);
        }
    }
}

File: java/src/main/java/io/cucumber/cucumberexpressions/ExpressionFactory.java
package io.cucumber.cucumberexpressions;

import org.apiguardian.api.API;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * Creates a {@link CucumberExpression} or {@link RegularExpression} from a {@link String}
 * using heuristics. This is particularly useful for languages that don't have a
 * literal syntax for regular expressions. In Java, a regular expression has to be represented as a String.
 *
 *  A string that starts with `^` and/or ends with `$` (or written in script style, i.e. starting with `/` 
 *  and ending with `/`) is considered a regular expression.
 *  Everything else is considered a Cucumber expression.
 */
@API(status = API.Status.STABLE)
public final class ExpressionFactory {

    private static final Pattern PARAMETER_PATTERN = Pattern.compile("((?:\\\\){0,2})\\{([^}]*)\\}");

    private final ParameterTypeRegistry parameterTypeRegistry;

    public ExpressionFactory(ParameterTypeRegistry parameterTypeRegistry) {
        this.parameterTypeRegistry = parameterTypeRegistry;
    }

    public Expression createExpression(String expressionString) {
        int length = expressionString.length();
        int lastCharIndex = 0;
        char firstChar = 0;
        char lastChar = 0;
        if (length > 0) {
            lastCharIndex = length - 1;
            firstChar = expressionString.charAt(0);
            lastChar = expressionString.charAt(lastCharIndex);
        }
        if (firstChar == '^' || lastChar == '$') {
            // java regexp => create from regexp
            return this.createRegularExpressionWithAnchors(expressionString);
        } else if (firstChar == '/' && lastChar == '/') {
            // script style regexp => create from regexp
            return new RegularExpression(Pattern.compile(expressionString.substring(1, lastCharIndex)), this.parameterTypeRegistry);
        } else {
            // otherwise create cucumber style expression
            return new CucumberExpression(expressionString, this.parameterTypeRegistry);
        }
    }

    private RegularExpression createRegularExpressionWithAnchors(String expressionString) {
        try {
            return new RegularExpression(Pattern.compile(expressionString), parameterTypeRegistry);
        } catch (PatternSyntaxException e) {
            if (PARAMETER_PATTERN.matcher(expressionString).find()) {
                throw new CucumberExpressionException("You cannot use anchors (^ or $) in Cucumber Expressions. Please remove them from " + expressionString, e);
            }
            throw e;
        }
    }
}
